README.md
# Flutter Utils

A set of useful tools to make development more easily.

[Source code](https://github.com/gabrimatic/FlutterUtils/)

[Pub.dev page](https://pub.dev/packages/gabrimatic_utils)

[Developer](https://gabrimatic.info)


CHANGELOG.md
## [0.0.7] - 2020/11/09

- Files structure updated
- TextArrow widget added
- TextIcon widget added


lib/string.dart
import 'package:easy_localization/easy_localization.dart';

extension GabrimaticStringUtils on String {
  /// Adds an extra 0 in behind a String
  /// It is useful when you want to show a time; It converts '5' to '05'
  String get gAddExtraZero => (this.length == 1) ? ('0' + this) : this;

  /// If you are using [easy_localization] package, to apply translation on strings, simply use [gTR] extension!
  /// example: 'someText'.gTR
  String get gTR => tr(this);

  /// The [isEmpty] method with professional check!
  bool get gIsEmpty {
    if (this == null || this.isEmpty || this == 'null') return true;
    String data = this;
    data = data.replaceAll(' ', '');
    data = data.replaceAll('\n', '');
    if (data == null || data.isEmpty)
      return true;
    else
      return false;
  }

  bool get gIsNotEmpty => !this.gIsEmpty;
}


lib/date_tools.dart
import 'package:easy_localization/easy_localization.dart';
import 'package:flutter/material.dart';
import 'package:persian_date/persian_date.dart';
import 'package:persian_datetime_picker/persian_datetime_picker.dart';

import 'string.dart';

/// This class uses [persian_date] and [persian_datetime_picker] packages for jalali dates.
class GabrimaticDateUtils {
  GabrimaticDateUtils({this.isJalali = false});

  bool isJalali;

  // Get Date times
  int nowTimeStamp({bool withTime = true}) =>
      nowDateTime(withTime: withTime).millisecondsSinceEpoch;

  DateTime nowDateTime({bool withTime = true}) => (withTime)
      ? DateTime.now()
      : DateTime.fromMillisecondsSinceEpoch(
          removeTimeFromDate(DateTime.now().millisecondsSinceEpoch));

  int removeTimeFromDate(int timestamp) {
    var date = DateTime.fromMillisecondsSinceEpoch(timestamp);
    return DateTime(date.year, date.month, date.day).millisecondsSinceEpoch;
  }

  /// Returns an String like: 'Tuesday, 25 Oct, 2020'
  String dateTimeToFancyString(DateTime dateTime) {
    if (isJalali) {
      var persianDate = PersianDate(gregorian: dateTime.toString());
      return (persianDate.weekdayname +
          '، ' +
          persianDate.day.toString() +
          ' ' +
          persianDate.monthname +
          ' ' +
          persianDate.year.toString());
    } else
      return DateFormat('EEEE, d MMM, yyyy').format(dateTime);
  }

  String getTimeFromDateTime(DateTime dateTime) =>
      dateTime.hour.toString().gAddExtraZero +
      ':' +
      dateTime.minute.toString().gAddExtraZero;

  String getTimeFromTimeOfDay(TimeOfDay timeOfDay) =>
      timeOfDay.hour.toString().gAddExtraZero +
      ':' +
      timeOfDay.minute.toString().gAddExtraZero;

  // Timestamps
  String readTimestamp(int timestamp, {bool withTime = true}) => (isJalali)
      ? _readJalaliTimestamp(timestamp, withTime)
      : _readGregorianTimestamp(timestamp, withTime);

  String _readGregorianTimestamp(int timestamp, bool withTime) {
    var date = DateTime.fromMillisecondsSinceEpoch(timestamp);
    return (withTime)
        ? (date.year.toString() +
            '/' +
            date.month.toString() +
            '/' +
            date.day.toString() +
            ', at ' +
            date.hour.toString().gAddExtraZero +
            ':' +
            date.minute.toString().gAddExtraZero)
        : (date.year.toString() +
            '/' +
            date.month.toString() +
            '/' +
            date.day.toString());
  }

  String _readJalaliTimestamp(int timestamp, bool withTime) {
    var persianDate = PersianDate(
        gregorian: DateTime.fromMillisecondsSinceEpoch(timestamp).toString());
    return (withTime)
        ? (persianDate.year.toString() +
            '/' +
            persianDate.month.toString() +
            '/' +
            persianDate.day.toString() +
            '، ' +
            persianDate.hour.toString().gAddExtraZero +
            ':' +
            persianDate.minute.toString().gAddExtraZero)
        : (persianDate.year.toString() +
            '/' +
            persianDate.month.toString() +
            '/' +
            persianDate.day.toString());
  }

  // Date pickers
  Future<DateTime> showDatePickerDateTime(BuildContext context,
          {DateTime initialDate,
          DateTime firstDate,
          DateTime lastDate}) async =>
      (isJalali)
          ? await _showJalaliDatePicker(
              context, initialDate, firstDate, lastDate)
          : await _showGregorianDatePicker(
              context, initialDate, firstDate, lastDate);

  Future<int> showDatePickerTimestamp(BuildContext context,
          {DateTime initialDate,
          DateTime firstDate,
          DateTime lastDate,
          bool forWeb}) async =>
      (await showDatePickerDateTime(context,
              initialDate: initialDate,
              firstDate: firstDate,
              lastDate: lastDate))
          .millisecondsSinceEpoch;

  Future<DateTime> _showGregorianDatePicker(BuildContext context,
      DateTime initialDate, DateTime firstDate, DateTime lastDate) async {
    var nowDate = DateTime.now();
    return (await showDatePicker(
        context: context,
        initialDate: initialDate ?? nowDate,
        firstDate:
            firstDate ?? DateTime(nowDate.year, nowDate.month, nowDate.day),
        lastDate: lastDate ?? DateTime(nowDate.year).add(Duration(days: 365)),
        builder: (ctx, child) => child));
  }

  Future<DateTime> _showJalaliDatePicker(BuildContext context,
      DateTime initialDate, DateTime firstDate, DateTime lastDate) async {
    var initialJalaliDate = PersianDate(
        gregorian: initialDate.toString().isEmpty
            ? DateTime.now().toString()
            : initialDate.toString());
    var firstJalaliDate = PersianDate(
        gregorian: firstDate.toString().isEmpty
            ? DateTime.now().toString()
            : initialDate.toString());
    var lastJalaliDate = PersianDate(
        gregorian:
            lastDate ?? DateTime.now().add(Duration(days: 365)).toString());

    String pickedDate;
    await showDialog(
        context: context,
        child: PersianDateTimePicker(
          initial:
              "${initialJalaliDate.year}/${initialJalaliDate.month}/${initialJalaliDate.day}",
          min:
              "${firstJalaliDate.year}/${firstJalaliDate.month}/${firstJalaliDate.day}",
          max:
              "${lastJalaliDate.year}/${lastJalaliDate.month}/${lastJalaliDate.day}",
          onSelect: (date) => pickedDate = date,
        ));

    if (pickedDate.gIsEmpty) return null;

    return PersianDate().jalaliToGregorian(pickedDate.replaceAll('/', '-'));
  }

  // Time picker
  Future<TimeOfDay> showTheTimePicker(BuildContext context,
      {bool use24HourFormat = true}) async {
    TimeOfDay picked = (await showTimePicker(
        context: context,
        initialTime: TimeOfDay.now(),
        builder: (BuildContext context, Widget child) => MediaQuery(
            data: MediaQuery.of(context)
                .copyWith(alwaysUse24HourFormat: use24HourFormat),
            child: child)));
    if (picked == null) return null;
    return picked;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:gabrimatic_utils/gabrimatic_utils.dart';

void main() {
  runApp(Creator(child: ExampleApp()));
}

class ExampleApp extends StatelessWidget {
  const ExampleApp({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        children: [
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextButton(
              onPressed: () {
                Creator.recreate(context);
              },
              child: Text(
                'Restart App!',
                style: Theme.of(context).textTheme.headline4,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextButton(
              onPressed: () {
                DateUtils(isJalali: false).showDatePickerDateTime(context);
              },
              child: Text(
                'Show datepicker',
                style: Theme.of(context).textTheme.headline4,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextButton(
              onPressed: () {
                DateUtils(isJalali: false).showTheTimePicker(context);
              },
              child: Text(
                'Show timepicker',
                style: Theme.of(context).textTheme.headline4,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextButton(
              onPressed: () {
                print(''.gIsEmpty); // returns: true
                print('null'.gIsEmpty); // returns: true
                print('    '.gIsEmpty); // returns: true
              },
              child: Text(
                'Check string',
                style: Theme.of(context).textTheme.headline4,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextButton(
              onPressed: () {
                print('5'.addExtraZero); // returns: 05
              },
              child: Text(
                'Add zero',
                style: Theme.of(context).textTheme.headline4,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextButton(
              onPressed: () {
                print(enDigitToFaDigit('123456789')); // returns: ٠۱۲۳۴۵۶۷۸۹
              },
              child: Text(
                'Convert english digits to farsi digits',
                style: Theme.of(context).textTheme.headline4,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextButton(
              onPressed: () {
                shortIf(
                  p1: 'First value', // First parameter in condition
                  // p2: 'Second value', if p2 was null, p1 returns
                  returnValue: 'First is Second',
                );
              },
              child: Text(
                'Short if',
                style: Theme.of(context).textTheme.headline4,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextButton(
              onPressed: () {
                exitApp();
              },
              child: Text(
                'Exit!',
                style: Theme.of(context).textTheme.headline4,
              ),
            ),
          ),
        ],
      ),
    );
  }
}


lib/gabrimatic/utils.dart
import 'package:easy_localization/easy_localization.dart';
import 'package:flutter/services.dart';

/// p1: first parameter in condition
/// p2: Second parameter in condition. default value is [Null]
/// returnValue: An object which returns if the condition was true
shortIf({
  dynamic p1,
  dynamic p2,
  dynamic returnValue,
}) =>
    (p1 == p2 || p2 == null) ? returnValue : p1;

void exitApp() => SystemChannels.platform.invokeMethod('SystemNavigator.pop');

///********** STRINGS **********/
/// Convert the farsi digits to english digits in a string
String faDigitToEnDigit(String digit) => (digit
    .replaceAll('٠', '0')
    .replaceAll('۱', '1')
    .replaceAll('۲', '2')
    .replaceAll('۳', '3')
    .replaceAll('۴', '4')
    .replaceAll('۵', '5')
    .replaceAll('۶', '6')
    .replaceAll('۷', '7')
    .replaceAll('۸', '8')
    .replaceAll('۹', '9'));

/// Convert the english digits to farsi digits in a string
String enDigitToFaDigit(String digit) => (digit
    .replaceAll('0', '٠')
    .replaceAll('1', '۱')
    .replaceAll('2', '۲')
    .replaceAll('3', '۳')
    .replaceAll('4', '۴')
    .replaceAll('5', '۵')
    .replaceAll('6', '۶')
    .replaceAll('7', '۷')
    .replaceAll('8', '۸')
    .replaceAll('9', '۹'));

extension StringUtils on String {
  /// Adds an extra 0 in behind a String
  /// It is useful when you want to show a time; It converts '5' to '05'
  String get addExtraZero => (this.length == 1) ? ('0' + this) : this;

  /// If you are using [easy_localization] package, to apply translation on strings, simply use [gTR] extension!
  /// example: 'someText'.gTR
  String get gTR => tr(this);

  /// The [isEmpty] method with professional check!
  bool get gIsEmpty {
    if (this == null || this.isEmpty || this == 'null') return true;
    String data = this;
    data = data.replaceAll(' ', '');
    data = data.replaceAll('\n', '');
    if (data == null || data.isEmpty)
      return true;
    else
      return false;
  }

  bool get gIsNotEmpty => !this.gIsEmpty;
}


lib/gabrimatic/widgets.dart
import 'package:flutter/material.dart';

class Creator extends StatefulWidget {
  final Widget child;

  Creator({this.child});

  @override
  _CreatorState createState() => _CreatorState();

  /// Add [Creator] widget as the root widget of your app and call [recreate] method to restart the app!
  static recreate(BuildContext context) =>
      context.findAncestorStateOfType<_CreatorState>().recreateApp();
}

class _CreatorState extends State<Creator> {
  var _key = UniqueKey();

  void recreateApp() => setState(() => _key = UniqueKey());

  @override
  Widget build(BuildContext context) =>
      Container(key: _key, child: widget.child);
}

class TextIcon extends StatelessWidget {
  final IconData icon;
  final String text;
  final String preText;
  final double fontSize;
  final Color color;
  final bool showIcon;
  final Function onTap;

  TextIcon(
      {@required this.icon,
      @required this.text,
      this.preText = '',
      this.color = const Color(0XFF828282),
      this.fontSize = 16,
      this.showIcon = true,
      this.onTap});

  @override
  Widget build(BuildContext context) => GestureDetector(
        onTap: onTap,
        child: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            if (showIcon)
              Icon(
                icon,
                color: color,
              ),
            if (showIcon)
              SizedBox(
                width: 2,
              ),
            Text(
              preText,
              style: TextStyle(fontSize: fontSize, color: color),
            ),
            SizedBox(
              width: 2,
            ),
            Text(
              text,
              style: TextStyle(fontSize: fontSize, color: color),
            )
          ],
        ),
      );
}

class TextArrow extends StatelessWidget {
  final String text1;
  final String text2;
  final IconData icon;
  final double fontSize;
  final Color color;

  TextArrow(
      {@required this.text1,
      @required this.text2,
      this.icon,
      this.color = const Color(0XFF828282),
      this.fontSize = 16});

  @override
  Widget build(BuildContext context) => Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          if (icon != null)
            Icon(
              icon,
              color: color,
            ),
          if (icon != null)
            SizedBox(
              width: 2,
            ),
          Text(
            text1,
            style: TextStyle(fontSize: fontSize, color: color),
          ),
          SizedBox(
            width: 3,
          ),
          Icon(
            Icons.arrow_right,
            color: color,
          ),
          SizedBox(
            width: 3,
          ),
          Text(
            text2,
            style: TextStyle(fontSize: fontSize, color: color),
          )
        ],
      );
}


lib/gabrimatic/date_utils.dart
import 'package:easy_localization/easy_localization.dart';
import 'package:flutter/material.dart';
import 'package:persian_date/persian_date.dart';
import 'package:persian_datetime_picker/persian_datetime_picker.dart';
import 'package:gabrimatic_utils/gabrimatic_utils.dart';

/// This class uses [persian_date] and [persian_datetime_picker] packages for jalali dates.
class DateUtils {
  DateUtils({this.isJalali = false});

  bool isJalali;

  // Get Date times
  int nowTimeStamp({bool withTime = true}) =>
      nowDateTime(withTime: withTime).millisecondsSinceEpoch;

  DateTime nowDateTime({bool withTime = true}) => (withTime)
      ? DateTime.now()
      : DateTime.fromMillisecondsSinceEpoch(
          removeTimeFromTimestamp(DateTime.now().millisecondsSinceEpoch));

  int removeTimeFromTimestamp(int timestamp) {
    var date = DateTime.fromMillisecondsSinceEpoch(timestamp);
    return DateTime(date.year, date.month, date.day).millisecondsSinceEpoch;
  }

  /// Returns an String like: 'Tuesday, 25 Oct, 2020'
  String dateTimeToFancyString(DateTime dateTime) {
    if (isJalali) {
      var persianDate = PersianDate(gregorian: dateTime.toString());
      return (persianDate.weekdayname +
          '، ' +
          persianDate.day.toString() +
          ' ' +
          persianDate.monthname +
          ' ' +
          persianDate.year.toString());
    } else
      return DateFormat('EEEE, d MMM, yyyy').format(dateTime);
  }

  String getTimeFromDateTime(DateTime dateTime) =>
      dateTime.hour.toString().addExtraZero +
      ':' +
      dateTime.minute.toString().addExtraZero;

  String getTimeFromTimeOfDay(TimeOfDay timeOfDay) =>
      timeOfDay.hour.toString().addExtraZero +
      ':' +
      timeOfDay.minute.toString().addExtraZero;

  String appGetTimeFromTimestamp(int time) {
    var dateTime = DateTime.fromMillisecondsSinceEpoch(time);
    return dateTime.hour.toString().addExtraZero +
        ':' +
        dateTime.minute.toString().addExtraZero;
  }

  // Timestamps
  String readTimestamp(int timestamp, {bool withTime = true}) => (isJalali)
      ? _readJalaliTimestamp(timestamp, withTime)
      : _readGregorianTimestamp(timestamp, withTime);

  String _readGregorianTimestamp(int timestamp, bool withTime) {
    var date = DateTime.fromMillisecondsSinceEpoch(timestamp);
    return (withTime)
        ? (date.year.toString() +
            '/' +
            date.month.toString() +
            '/' +
            date.day.toString() +
            ', at ' +
            date.hour.toString().addExtraZero +
            ':' +
            date.minute.toString().addExtraZero)
        : (date.year.toString() +
            '/' +
            date.month.toString() +
            '/' +
            date.day.toString());
  }

  String _readJalaliTimestamp(int timestamp, bool withTime) {
    var persianDate = PersianDate(
        gregorian: DateTime.fromMillisecondsSinceEpoch(timestamp).toString());
    return (withTime)
        ? (persianDate.year.toString() +
            '/' +
            persianDate.month.toString() +
            '/' +
            persianDate.day.toString() +
            '، ' +
            persianDate.hour.toString().addExtraZero +
            ':' +
            persianDate.minute.toString().addExtraZero)
        : (persianDate.year.toString() +
            '/' +
            persianDate.month.toString() +
            '/' +
            persianDate.day.toString());
  }

  // Date pickers
  Future<DateTime> showDatePickerDateTime(BuildContext context,
          {DateTime initialDate,
          DateTime firstDate,
          DateTime lastDate}) async =>
      (isJalali)
          ? await _showJalaliDatePicker(
              context, initialDate, firstDate, lastDate)
          : await _showGregorianDatePicker(
              context, initialDate, firstDate, lastDate);

  Future<int> showDatePickerTimestamp(BuildContext context,
          {DateTime initialDate,
          DateTime firstDate,
          DateTime lastDate,
          bool forWeb}) async =>
      (await showDatePickerDateTime(context,
              initialDate: initialDate,
              firstDate: firstDate,
              lastDate: lastDate))
          .millisecondsSinceEpoch;

  Future<DateTime> _showGregorianDatePicker(BuildContext context,
      DateTime initialDate, DateTime firstDate, DateTime lastDate) async {
    var nowDate = DateTime.now();
    return (await showDatePicker(
        context: context,
        initialDate: initialDate ?? nowDate,
        firstDate:
            firstDate ?? DateTime(nowDate.year, nowDate.month, nowDate.day),
        lastDate: lastDate ?? DateTime(nowDate.year).add(Duration(days: 365)),
        builder: (ctx, child) => child));
  }

  Future<DateTime> _showJalaliDatePicker(BuildContext context,
      DateTime initialDate, DateTime firstDate, DateTime lastDate) async {
    var initialJalaliDate = PersianDate(
        gregorian: initialDate.toString().isEmpty
            ? DateTime.now().toString()
            : initialDate.toString());
    var firstJalaliDate = PersianDate(
        gregorian: firstDate.toString().isEmpty
            ? DateTime.now().toString()
            : initialDate.toString());
    var lastJalaliDate = PersianDate(
        gregorian:
            lastDate ?? DateTime.now().add(Duration(days: 365)).toString());

    String pickedDate;
    await showDialog(
        context: context,
        child: PersianDateTimePicker(
          initial:
              "${initialJalaliDate.year}/${initialJalaliDate.month}/${initialJalaliDate.day}",
          min:
              "${firstJalaliDate.year}/${firstJalaliDate.month}/${firstJalaliDate.day}",
          max:
              "${lastJalaliDate.year}/${lastJalaliDate.month}/${lastJalaliDate.day}",
          onSelect: (date) => pickedDate = date,
        ));

    if (pickedDate.gIsEmpty) return null;

    return PersianDate().jalaliToGregorian(pickedDate.replaceAll('/', '-'));
  }

  // Time picker
  Future<TimeOfDay> showTheTimePicker(BuildContext context,
      {bool use24HourFormat = true}) async {
    TimeOfDay picked = (await showTimePicker(
        context: context,
        initialTime: TimeOfDay.now(),
        builder: (BuildContext context, Widget child) => MediaQuery(
            data: MediaQuery.of(context)
                .copyWith(alwaysUse24HourFormat: use24HourFormat),
            child: child)));
    if (picked == null) return null;
    return picked;
  }
}
